ts/src/client.ts

```diff
@@ -0,0 +1,311 @@
+import {Metadata} from "./metadata";
+import {ChunkParser, Chunk, ChunkType} from "./ChunkParser";
+import {Code, httpStatusToCode} from "./Code";
+import {debug} from "./debug";
+import detach from "./detach";
+import {Transport, TransportConstructor, DefaultTransportFactory} from "./transports/Transport";
+import {MethodDefinition} from "./service";
+import {frameRequest} from "./util";
+import {ProtobufMessage} from "./message";
+
+export interface ClientRpcOptions {
+  host: string;
+  transport?: TransportConstructor;
+  debug?: boolean;
+}
+
+export interface Client<TRequest extends ProtobufMessage, TResponse extends ProtobufMessage> {
+  start(metadata?: Metadata.ConstructorArg): void;
+  send(message: TRequest): void;
+  finishSend(): void;
+  close(): void;
+
+  onHeaders(callback: (headers: Metadata) => void): void;
+  onMessage(callback: (message: TResponse) => void): void;
+  onEnd(callback: (code: Code, message: string, trailers: Metadata) => void): void;
+}
+
+export function client<TRequest extends ProtobufMessage, TResponse extends ProtobufMessage, M extends MethodDefinition<TRequest, TResponse>>(methodDescriptor: M, props: ClientRpcOptions): Client<TRequest, TResponse> {
+  return new ClientImpl<TRequest, TResponse, M>(methodDescriptor, props);
+}
+
+class ClientImpl<TRequest extends ProtobufMessage, TResponse extends ProtobufMessage, M extends MethodDefinition<TRequest, TResponse>> {
+  methodDefinition: M;
+  props: ClientRpcOptions;
+
+  started: boolean = false;
+  sentFirstMessage: boolean = false;
+  completed: boolean = false;
+  closed: boolean = false;
+  finishedSend: boolean = false;
+
+  onHeadersCallbacks: Array<(headers: Metadata) => void> = [];
```

Comment:
Pulling this into a middleware in the future might make more sense, the client then only has one callback for each life cycle method.

Right now we also don't have a way to remove callbacks that may already have been registered. I think this fine as long as we know it's a limitation